[V1] guidance backend for structured output + `auto` fallback mode

russellb opened the pull request:
This is the V1 integration for
[guidance](https://github.com/guidance-ai/llguidance) as a backend for
structured output. There is a V0 integration in #14589.

Since this is the second backend supported by V1, this PR also includes a new backend mode called `auto`, which provides opinionated behavior with fallback. This is not the default since the behavior may vary release to release.

This backend provides some key benefits to V1:

* Broader jsonschema support
* Quick startup performance for large schemas

Instead of precomputing the masks for all states, this is done on
the fly. We see very fast request startup times, even for large
schemas.

This should make V1 roughly feature equivalent to V0 in terms of the
types of schemas it can support.

More technical details are available in the llguidance git repo.

Signed-off-by: Russell Bryant <rbryant@redhat.com>
Co-authored-by: Loc Huynh <jc1da.3011@gmail.com>

russellb committed:
[V1] guidance backend and auto mode for structured output

This is the V1 integration for
[guidance](https://github.com/guidance-ai/llguidance) as a backend for
structured output. There is a V0 integration in #14589.

This backend provides some key benefits to V1:

* Broader jsonschema support
* Quick startup performance for large schemas

Instead of precomputing the masks for all states, this is done on
the fly. We see very fast request startup times, even for large
schemas.

This should make V1 roughly feature equivalent to V0 in terms of the
types of schemas it can support.

An `auto` mode is also included, which includes opinionated fallback
behavior based on our current understanding for varying feature support
and performance characteristics for different scenarios.

More technical details are available in the llguidance git repo.

Signed-off-by: Russell Bryant <rbryant@redhat.com>
Co-authored-by: Loc Huynh <jc1da.3011@gmail.com>
Co-authored-by: Michal Moskal <michal@moskal.me>

WoosukKwon commented:
@russellb I removed this from the v0.8.0 blockers

mmoskal commented:
llguidance 0.7.1 now has rollback() and fast-forward() APIs similar to xgrammar; it also uses new simpler LLMatcher interface instead of LLInterpreter; will try to adapt this PR later today

Related v0 updates:

https://github.com/russellb/vllm/compare/llguidance-v0-integration...mmoskal:llg_v0_matcher

russellb commented:
> llguidance 0.7.1 now has rollback() and fast-forward() APIs similar to xgrammar; it also uses new simpler LLMatcher interface instead of LLInterpreter; will try to adapt this PR later today
> 
> Related v0 updates:
> 
> [russellb/vllm@llguidance-v0-integration...mmoskal:llg_v0_matcher](https://github.com/russellb/vllm/compare/llguidance-v0-integration...mmoskal:llg_v0_matcher)

thanks!

I'm also turning on test coverage. I forgot to do that earlier.

mgoin reviewed (commented):
Failing v1 tests look related, PTAL

russellb commented:
> Failing v1 tests look related, PTAL

Thanks, @mgoin.

@mmoskal did you say you were already looking into this?

```
[2025-03-20T17:30:00Z] ERROR 03-20 17:30:00 [core.py:340] llguidance.gbnf_to_lark.ParseError: Expected name at line 2, '\nroot ::= select' ^ '_statement\nselect_st'
  |  
```

I'm going to put this in draft state while we get this resolved.

russellb commented:
> Oh, please just bump llguidance to latest.

Thanks, done! Let's see what CI says now ...

russellb commented:
@mmoskal CI is still failing. Can you take a look?  For example: https://buildkite.com/vllm/ci/builds/16023#0195b9b6-c23c-45d7-abbe-6e8968bb0aaa/207-5799

```
...

[2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]   File "/opt/venv/lib/python3.12/site-packages/llguidance/gbnf_to_lark.py", line 528, in any_to_lark
--
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]     return gbnf_to_lark(text)
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]            ^^^^^^^^^^^^^^^^^^
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]   File "/opt/venv/lib/python3.12/site-packages/llguidance/gbnf_to_lark.py", line 498, in gbnf_to_lark
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]     rules = parser.parse(text)
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]             ^^^^^^^^^^^^^^^^^^
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]   File "/opt/venv/lib/python3.12/site-packages/llguidance/gbnf_to_lark.py", line 193, in parse
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]     rule, pos = self._parse_rule(pos)
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]                 ^^^^^^^^^^^^^^^^^^^^^
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]   File "/opt/venv/lib/python3.12/site-packages/llguidance/gbnf_to_lark.py", line 326, in _parse_rule
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343]     raise ParseError(pos, "Expected ::=")
  | [2025-03-21T18:29:12Z] ERROR 03-21 18:29:12 [core.py:343] llguidance.gbnf_to_lark.ParseError: Expected ::= at line 1, 'not ' ^ 'a grammar'


```

mmoskal commented:
I can validate the request in `Processor._validate_structured_output` ; will get to that on Monday

However, I think we should also have some way of propagating errors from fill_mask/consume_token all the way to the user. In particular llguidance has a host of parameters that serve as "deterministic timeouts" to abort grammars that are too complex (the idea is that mask computations shouldn't take longer than ~20ms and grammar compilation longer than ~100ms). Such "timeouts" should be sent back to the user.

We never hit these in the 10k schemas in [maskbench](https://github.com/guidance-ai/jsonschemabench/tree/maskbench/maskbench) ; but I imagine some users may trigger them. 

Something for the next PR I guess.

russellb commented:
Great feedback. I agree we should improve error handling and propagation.

russellb commented:
@mmoskal I just realized we're using xgrammar for applying the bitmask since we didn't touch the GPU worker code in this PR. It's obviously working, but seems like we should use llguidance for consistency, even if it's just in a follow-up PR.

mmoskal commented:
It’s exactly the same mask format. Given that llg and xgr come up with it independently, maybe we just call it universal?

mmoskal commented:
This way you could on future apply it directly in softmax kernel.

aarnphm reviewed (approved):
Let's get this in. LGTM

edit: oh i didn't see the CI failure. We must fix that.

russellb committed:
Remove duplicate llguidance entry in requirements

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Allow list or dict with json_object for xgrammar

This is a bug, but we can't properly fix it until the next xgrammar
release, so just let the test be a little bit more flexible for now.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
The last CI failure was actually a bug with xgrammar, but we need an xgrammar release to fix it properly. I put in a change in the tests to make it a bit less strict until we can fix the issue.

WoosukKwon reviewed (commented):
@aarnphm Could you please take a look?

simon-mo commented:
merging for release
